engine/src/execution_graph/logic_controllers/taskflow/graph.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+#include "kernel.h"
+#include "kpair.h"
+
+namespace ral {
+namespace cache {
+
+
+static std::shared_ptr<ral::cache::CacheMachine> create_cache_machine(const cache_settings& config) {
+	std::shared_ptr<ral::cache::CacheMachine> machine;
+	if (config.type == CacheType::NON_WAITING) {
+		machine =  std::make_shared<ral::cache::NonWaitingCacheMachine>();
+	}
+	else if (config.type == CacheType::SIMPLE or config.type == CacheType::FOR_EACH) {
+		machine =  std::make_shared<ral::cache::CacheMachine>();
+	} else if (config.type == CacheType::CONCATENATING) {
+		machine =  std::make_shared<ral::cache::ConcatenatingCacheMachine>();
+	}
+	return machine;
+}
+
+static std::vector<std::shared_ptr<ral::cache::CacheMachine>> create_cache_machines(const cache_settings& config) {
+	std::vector<std::shared_ptr<ral::cache::CacheMachine>> machines;
+	for (size_t i = 0; i < config.num_partitions; i++) {
+		machines.push_back(create_cache_machine(config));
+	}
+	return machines;
+}
+
+class graph {
+protected:
+	struct Edge {
+		std::int32_t source;
+		std::int32_t target;
+		std::string source_port_name;
+		std::string target_port_name;
+
+		bool operator<(const Edge & e) const { return this->target < e.target; }
+		bool operator==(const Edge & e) const { return this->target == e.target; }
+	};
+
+public:
+	graph() {
+		container_[head_id_] = nullptr;	 // sentinel node
+	}
+	graph(const graph &) = default;
+	graph & operator=(const graph &) = default;
+
+	kpair operator+=(kpair p) {
+		std::string source_port_name = std::to_string(p.src->get_id());
+		std::string target_port_name = std::to_string(p.dst->get_id());
+
+		if(p.has_custom_source()) {
+			source_port_name = p.src_port_name;
+		}
+		if(p.has_custom_target()) {
+			target_port_name = p.dst_port_name;
+		}
+		this->add_edge(p.src, p.dst, source_port_name, target_port_name, p.cache_machine_config);
+		return p;
+	}
+
+	void check_and_complete_work_flow() {
+		for(auto node : container_) {
+			kernel * kernel_node = node.second;
+			if(kernel_node) {
+				if(get_neighbours(kernel_node).size() == 0) {
+					Edge fake_edge = {.source = (int32_t) kernel_node->get_id(),
+						.target = -1,
+						.source_port_name = std::to_string(kernel_node->get_id()),
+						.target_port_name = ""};
+					edges_[kernel_node->get_id()].insert(fake_edge);
+				}
+			}
+		}
+	}
+
+	void execute() {
+		check_and_complete_work_flow();
+
+		std::vector<BlazingThread> threads;
+		std::set<std::pair<size_t, size_t>> visited;
+		std::deque<size_t> Q;
```

Comment:
what the fuck is Q? This is not a very appropriate name.

Reply:
all right, I will change in next PR.